core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1967,6 +1967,25 @@ private[spark] object Utils extends Logging {
     }
   }
 
+  /**
+   * To figure out which shutdown hook makes execution slow, a logWrapper for shutdown hook
+   * is used to record execute time for each shutdown hook.
+   *
+   * @param f
+   * @param nextHook
+   * @tparam T
+   */
+  def shutdownHookLogWrapper[T](f: => T, nextHook: SparkShutdownHook): Unit = {
+    val startTime = System.currentTimeMillis()
+    try {
+      logInfo(s"Shutdown hook ${nextHook.caller} starts!")
```

Comment:
Nit: this doesn't need to be in the try-catch, and remove the exclamation, I think

Reply:
I think it's possible the function wrapped will throw exception so it's still necessary to try-catch-finally